### PR TITLE
Add shutdown tests [ECR-850], [ECR-847]

### PR DESCRIPTION
**Pull request by alekseysidorov**
- Fix "can't cancel routine" error.
 - Add test for shutdown request.
 - Add test for a possible regression in "create_genesis_block" behaviour.

### CODE REVIEW

File: exonum/src/events/network.rs [hunk 1]
```diff
@@ -311,11 +312,6 @@ impl RequestHandler {
                         let fut = cancel_sender
                             .take()
                             .ok_or_else(|| other_error("shutdown twice"))
```
Comment (vitvakatu) [inline exonum/src/events/network.rs]: Is it real scenario? I mean, with this change, cancel_handler will return `Ok` in any situation, won't it?

Comment (alekseysidorov) [reply exonum/src/events/network.rs]: It seems to be no way to do it twice.

File: exonum/src/blockchain/mod.rs [hunk 2]
```diff
@@ -82,8 +82,8 @@ pub struct Blockchain {
 
 impl Blockchain {
     /// Constructs a blockchain for the given `storage` and list of `services`.
-    pub fn new(
-        storage: Box<Database>,
+    pub fn new<D: Into<Arc<Database>>>(
+        storage: D,
```
Comment (DarkEld3r) [inline exonum/src/blockchain/mod.rs]: Technically, this is a breaking change, right?

Comment (alekseysidorov) [reply exonum/src/blockchain/mod.rs]: I do not think so. In old code you can pass `Box::new(MemoryDb::new))` but now you need to convert `MemoryDb` into `Arc<Database>` manually or just pass `MemoryDb::new` I you do not need customization.

**Comment by matklad** [comment]
Do we have a tests where a shutdown message is send twice? Should we have one?

**Pull request merged**